File: backend/routers/tasks.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from typing import List

from database import get_db, Task, Employee
from schemas import TaskCreate, TaskOut, TaskUpdate
from core import get_current_user
from routers.audit import log_action

router = APIRouter(prefix="/tasks", tags=["tasks"])

@router.post("/", response_model=TaskOut)
async def create_task(
    task: TaskCreate,
    current_user: Employee = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    if current_user.role != "admin":
        raise HTTPException(status_code=403, detail="Faqat admin vazifa yarata oladi")
    
    new_task = Task(
        title=task.title,
        description=task.description,
        status=task.status,
        assigned_to=task.assigned_to,
        created_by=current_user.id,
        due_date=task.due_date
    )
    db.add(new_task)
    
    await log_action(db, current_user.id, "YANGI_VAZIFA", f"Vazifa: {new_task.title}. Assigned to ID: {new_task.assigned_to}")
    
    await db.commit()
    await db.refresh(new_task)

    # Telegram orqali xaridorni yoki mas'ulni xabardor qilish
    try:
        from bot import bot
        # Mas'ul xodim ma'lumotlarini olish
        result = await db.execute(select(Employee).where(Employee.id == task.assigned_to))
        assigned_emp = result.scalars().first()
        
        if assigned_emp and assigned_emp.telegram_id:
            msg = (
                f"📝 <b>Yangi Vazifa!</b>\n\n"
                f"📌 <b>Nomi:</b> {new_task.title}\n"
                f"📄 <b>Izoh:</b> {new_task.description or '-'}\n"
                f"👤 <b>Kimdan:</b> {current_user.full_name or current_user.username}\n"
                f"📅 <b>Muddat:</b> {new_task.due_date.strftime('%d.%m.%Y') if new_task.due_date else '-'}\n\n"
                f"<i>Vazifani bajarish uchun dasturga kiring.</i>"
            )
            await bot.send_message(assigned_emp.telegram_id, msg, parse_mode="HTML")
    except Exception as e:
        print(f"Telegram notification error: {e}")

    return new_task

@router.get("/", response_model=List[TaskOut])
async def get_tasks(
    current_user: Employee = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    # Admins/Managers see all tasks or tasks they created? Let's say all for now.
    # Employees see only their assigned tasks.
    
    query = select(Task)
    
    if current_user.role != "admin":
        query = query.where(Task.assigned_to == current_user.id)
        
    result = await db.execute(query)
    return result.scalars().all()

@router.put("/{task_id}", response_model=TaskOut)
async def update_task(
    task_id: int,
    task_update: TaskUpdate,
    current_user: Employee = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    result = await db.execute(select(Task).where(Task.id == task_id))
    task = result.scalars().first()
    
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")

    # Permission check: 
    # Admin/Manager can update anything.
    # Worker can only update status of their own task.
    
    if current_user.role != "admin":
        if task.assigned_to != current_user.id:
            raise HTTPException(status_code=403, detail="Faqat o'zingizga biriktirilgan vazifa statusini o'zgartira olasiz")
        # Workers can only update status
        if task_update.title or task_update.description or task_update.assigned_to:
             raise HTTPException(status_code=403, detail="Workers can only update status")
    
    for key, value in task_update.dict(exclude_unset=True).items():
        setattr(task, key, value)
    
    await log_action(db, current_user.id, "VAZIFA_YANGILANDI", f"Vazifa ID: {task_id}. Status: {task.status}")
    
    await db.commit()
    await db.refresh(task)
    return task

@router.delete("/{task_id}")
async def delete_task(
    task_id: int,
    current_user: Employee = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    if current_user.role != "admin":
        raise HTTPException(status_code=403, detail="Faqat admin vazifani o'chira oladi")
        
    result = await db.execute(select(Task).where(Task.id == task_id))
    task = result.scalars().first()
    
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
        
    await db.delete(task)
    
    await log_action(db, current_user.id, "VAZIFA_OCHIRILDI", f"Vazifa ID: {task_id}. Title: {task.title}")
    
    await db.commit()
    return {"message": "Task deleted"}

File: php/app/routes/tasks.php
<?php
/**
 * /tasks — xodim vazifalari (faqat admin yaratadi/o'chiradi).
 *
 * Interfeysda bu "Xodimlar" sahifasi ichida ko'rinadi.
 * Xodimning o'zi faqat O'Z vazifasining holatini o'zgartira oladi.
 */

declare(strict_types=1);

// =======================================================================
// POST /tasks/
// =======================================================================
Router::post('/', function (): never {
    $user = Auth::require(['admin'], 'Faqat admin vazifa yarata oladi');
    $b = Http::body();

    $title = Http::reqStr($b, 'title', 300, 'Nomi');
    $assignedTo = Http::reqInt($b, 'assigned_to', "Mas'ul xodim");

    if (Db::one('SELECT id FROM employees WHERE id = ?', [$assignedTo]) === null) {
        fail(404, 'Xodim topilmadi');
    }

    $data = [
        'title'       => $title,
        'description' => Http::str($b, 'description', null, 2000, 'Izoh'),
        'status'      => Http::enum($b, 'status', ['pending', 'in_progress', 'completed'], 'pending', 'Holat'),
        'assigned_to' => $assignedTo,
        'created_by'  => (int)$user['id'],
        'created_at'  => Tz::now(),
        'due_date'    => Tz::parseFilterDate(Http::str($b, 'due_date', null, 40)),
    ];

    $id = Db::tx(function () use ($data, $user, $title, $assignedTo): int {
        $id = Db::insert('tasks', $data);
        Audit::log((int)$user['id'], 'YANGI_VAZIFA', "Vazifa: $title. Assigned to ID: $assignedTo");
        return $id;
    });

    // Eslatma: Python versiyasi bu yerda Telegram orqali xabar yuborardi.
    // Shared hosting'da bot ishlamaydi (uzluksiz jarayon yo'q), shuning
    // uchun xabar yuborilmaydi. Vazifaning o'zi normal yaratiladi.

    Http::json(Shape::task(Db::one('SELECT * FROM tasks WHERE id = ?', [$id])));
});

// =======================================================================
// GET /tasks/
// =======================================================================
Router::get('/', function (): never {
    $user = Auth::user();

    // Admin hammasini ko'radi, qolganlar faqat o'zinikini.
    $rows = $user['role'] === 'admin'
        ? Db::all('SELECT * FROM tasks ORDER BY id DESC LIMIT 1000')
        : Db::all('SELECT * FROM tasks WHERE assigned_to = ? ORDER BY id DESC LIMIT 1000', [(int)$user['id']]);

    Http::json(array_map(fn($r) => Shape::task($r), $rows));
});

// =======================================================================
// PUT /tasks/{task_id}
// =======================================================================
Router::put('/{task_id}', function (array $p): never {
    $user = Auth::user();
    $taskId = Router::id($p, 'task_id');
    $b = Http::body();

    $task = Db::one('SELECT * FROM tasks WHERE id = ?', [$taskId]);
    if ($task === null) {
        fail(404, 'Task not found');
    }

    $isAdmin = $user['role'] === 'admin';

    if (!$isAdmin) {
        if ((int)$task['assigned_to'] !== (int)$user['id']) {
            fail(403, "Faqat o'zingizga biriktirilgan vazifa statusini o'zgartira olasiz");
        }
        // Xodim faqat HOLATNI o'zgartira oladi.
        if (!empty($b['title']) || !empty($b['description']) || !empty($b['assigned_to'])) {
            fail(403, 'Workers can only update status');
        }
    }

    $data = [];
    if (array_key_exists('status', $b)) {
        $data['status'] = Http::enum($b, 'status', ['pending', 'in_progress', 'completed'], null, 'Holat');
    }
    if ($isAdmin) {
        if (array_key_exists('title', $b)) {
            $data['title'] = Http::reqStr($b, 'title', 300, 'Nomi');
        }
        if (array_key_exists('description', $b)) {
            $data['description'] = Http::str($b, 'description', null, 2000, 'Izoh');
        }
        if (array_key_exists('assigned_to', $b)) {
            $data['assigned_to'] = Http::reqInt($b, 'assigned_to', "Mas'ul xodim");
        }
        if (array_key_exists('due_date', $b)) {
            $data['due_date'] = Tz::parseFilterDate(Http::str($b, 'due_date', null, 40));
        }
    }

    Db::tx(function () use ($data, $taskId, $user): void {
        if ($data !== []) {
            Db::update('tasks', $taskId, $data);
        }
        $now = Db::one('SELECT status FROM tasks WHERE id = ?', [$taskId]);
        Audit::log((int)$user['id'], 'VAZIFA_YANGILANDI',
            "Vazifa ID: $taskId. Status: " . ($now['status'] ?? '-'));
    });

    Http::json(Shape::task(Db::one('SELECT * FROM tasks WHERE id = ?', [$taskId])));
});

// =======================================================================
// DELETE /tasks/{task_id}
// =======================================================================
Router::delete('/{task_id}', function (array $p): never {
    $user = Auth::require(['admin'], "Faqat admin vazifani o'chira oladi");
    $taskId = Router::id($p, 'task_id');

    $task = Db::one('SELECT * FROM tasks WHERE id = ?', [$taskId]);
    if ($task === null) {
        fail(404, 'Task not found');
    }

    Db::tx(function () use ($taskId, $user, $task): void {
        Db::delete('tasks', $taskId);
        Audit::log((int)$user['id'], 'VAZIFA_OCHIRILDI',
            "Vazifa ID: $taskId. Title: {$task['title']}");
    });

    Http::json(['message' => 'Task deleted']);
});
